### ISSUE TITLE
Add a Conjugate Gradient solver

### ISSUE THREAD

**Issue by hxu296** [pypose/pypose#311]
### Motivation: sparse LM optimizer relies on a sparse Ax = b solver
Hi! We are working on a sparse Levenberg–Marquardt optimizer, and we have already sparsified the Jacobian matrix and A matrix (derived from `J_T @ J`) into COO tensors. By default, the pypose LM optimizer uses Cholesky solver to [solve for the the param update for LM](https://github.com/pypose/pypose/blob/6639519647f12b83e532d31f1092a647f7895c51/pypose/optim/optimizer.py#L556), which does not support sparse PyTorch tensors.

### We tried to implement a sparse CG solver ourselves, but it did not work very well
Therefore, we implemented a [Conjugate Gradient solver](https://github.com/pypose/pypose/blob/6639519647f12b83e532d31f1092a647f7895c51/pypose/optim/solver.py#L214) where the A matrix can be sparsed. We reference [this open source project](https://github.com/nicknytko/numml/blob/main/numml/krylov.py). This solver works for optimizing BAL problems - we can see the loss decreasing - but the terminal loss is still very high compared to GTSAM's LM and Scipy's non-linear least square optimizer (which used trf optimizer):

![image](https://github.com/pypose/pypose/assets/61036578/ae5a9f5b-7b6d-4e2c-86be-9441de7b449b)
**Figure: Our terminal least square loss with CG solver**

![image](https://github.com/pypose/pypose/assets/61036578/55da1d21-6807-47a8-b56a-9927ac61d948)
**Figure: GTSAM's least square loss on the same dataset**

### Experiment shows that scipy's CG solver can achieve better end-to-end result than our CG solver
Going back to the CG solver. We suspected that the loss difference is due to the inaccuracy of our CG solver. Luckily, scipy also has official CG solver implementation, so we compared our CG with scipy's CG. It turns out, with scipy's CG solver we achieved much better least square loss, so our original CG solver may be doing something wrong here...
![image](https://github.com/pypose/pypose/assets/61036578/721b6ec6-9c20-456c-ad2f-ff5c5734d3a0)
**Figure: Our LM using scipy's CG solver**

### Towards a more accurate CG solver
Therefore, we came to the conclusion that we need a more accurate sparse linear system solver. For example, [JAX hand-crafted its CG solver](https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.sparse.linalg.cg.html) that is bit-accurate with scipy.

![image](https://github.com/pypose/pypose/assets/61036578/73f37fc0-6401-46bb-8d96-02b35fc29632)

**Figure: JAX's CG solver**

### However, we could really use some help for the sparse CG solver from the pypose community :)

I am Huan Xu in the pypose Slack, shoot me a message if you want to collaborate with us on this... Thanks!

**Comment by hxu296** [comment]
solved by #312
